### PR TITLE
Add a profiler lock protecting the profile region name stack

### LINKED ISSUE #3520: test_multi_app.sh presents region exit without entry error when an old controller instance is still running
**Describe the bug**
I tried to run test_multi_app.sh and I expected it to consistently pass instead it rarely fails.

**GEOPM version**
v3.1

**Expected behavior**
With test_mult_app.sh configured not to produce trace files, and with GEOPM_CTL_LOCAL present:
```
bgeltz@x4709c7s0b0n0:~/output> PATH=${HOME}/build/stress-ng/bin:${PATH} mpiexec -n 1 -ppn 1 ~/geopm/integration/test/test_multi_app.sh
/home/bgeltz/output
stress-ng: info:  [204962] setting to a 10 secs run per stressor
stress-ng: info:  [204962] dispatching hogs: 1 cpu
stress-ng: info:  [204962] skipped: 0
stress-ng: info:  [204962] passed: 1: cpu (1)
stress-ng: info:  [204962] failed: 0
stress-ng: info:  [204962] metrics untrustworthy: 0
stress-ng: info:  [204962] successful run completed in 10.00 secs
Beginning loop of 2 iterations.
Completed loop.

bgeltz@x4709c7s0b0n0:~/output> ls
temp_config.json  test_multi_app_report.yaml-x4709c7s0b0n0
```

**Actual behavior**
Exit before entry error, and no report files output:
```
bgeltz@x4709c7s0b0n0:~/output> PATH=${HOME}/build/stress-ng/bin:${PATH} mpiexec -n 1 -ppn 1 ~/geopm/integration/test/test_multi_app.sh
/home/bgeltz/output
stress-ng: info:  [204894] setting to a 10 secs run per stressor
stress-ng: info:  [204894] dispatching hogs: 1 cpu
Error: <geopm> Invalid argument: ValidateRecord::filter(): Region exit without entry Region=0x00000000: at src/ValidateRecord.cpp:70
Error: <geopm> Invalid argument: ValidateRecord::filter(): Region exit without entry Region=0x00000000: at src/ValidateRecord.cpp:70
stress-ng: info:  [204894] skipped: 0
stress-ng: info:  [204894] passed: 1: cpu (1)
stress-ng: info:  [204894] failed: 0
stress-ng: info:  [204894] metrics untrustworthy: 0
stress-ng: info:  [204894] successful run completed in 10.00 secs
Beginning loop of 2 iterations.
Completed loop.

bgeltz@x4709c7s0b0n0:~/output> ls
temp_config.json
```

**Additional context**
The above runs were done back-to-back with no changes in between.

### PR DESCRIPTION
**Pull request by cmcantalupo**
- Fixes #3520

### CODE REVIEW

File: integration/test/test_multi_app.sh [hunk 1]
```diff
@@ -25,6 +25,22 @@ GEOPM_REPORT_SIGNALS=TIME@package \
...
+# Check that second controller startup fails with correct error message
+TEMP_LOG=$(mktemp)
+geopmctl &> ${TEMP_LOG}
+if [ $? -eq 0 ]; then
+    echo "Error: Attempt to create second GEOPM Controller succeeded" 1>&2
+    cat ${TEMP_LOG}
+    rm ${TEMP_LOG}
+    exit -1
+fi
+if ! grep -q "requested control of application profile, but the geopm service already has a application profile controlling client" ${TEMP_LOG}; then
```
Comment (bgeltz) [inline integration/test/test_multi_app.sh]:
```suggestion
if ! grep -q "requested control of application profile, but the geopm service already has an application profile controlling client" ${TEMP_LOG}; then
```

File: libgeopmd/src/LevelZeroIOGroup.cpp [hunk 2]
```diff
@@ -811,6 +811,10 @@ namespace geopm
             for (int domain_idx = 0; domain_idx < m_platform_topo.num_domain(
                                      signal_domain_type(sv.first)); ++domain_idx) {
                 try {
+                    if (sv.first.find("_PERFORMANCE_") != std::string::npos && !m_is_perf_factor_enabled) {
+                        unsupported_signal_names.push_back(sv.first);
+                        break;
+                    }
```
Comment (bgeltz) [inline libgeopmd/src/LevelZeroIOGroup.cpp:817]: What's this change doing here?  What does it fix re: this bug?

**Review by bgeltz — CHANGES_REQUESTED** [review]

File: libgeopmd/src/LevelZeroIOGroup.cpp (hunk 2 above)
Comment (cmcantalupo) [reply libgeopmd/src/LevelZeroIOGroup.cpp:817]: It is unrelated.  @dannosliwcd sited the issue.  I can open a separate pull pull request it needed.

Comment (cmcantalupo) [reply libgeopmd/src/LevelZeroIOGroup.cpp:817]: I have updated the commit message just now.
https://github.com/geopm/geopm/pull/3580/commits/1c18bd84ebe12f9522d1214f61acfe3885c9f705

Comment (dannosliwcd) [reply libgeopmd/src/LevelZeroIOGroup.cpp:817]: I didn't observe any problematic behavior, just found this through unplanned code inspection.

I was digging in this code while looking into adding memory bandwidth signal and found it odd that we check whether the signal's name matches name patterns of unsupported signals _after_ we attempt to read the signal (`sv.second.m_devpool_func(domain_idx);`).

So with the prior implementation, we try to read the signal, then we check whether it's a signal that we know we should avoid reading.

File: geopmdpy/geopmdpy/system_files.py [hunk 3]
```diff
@@ -1277,7 +1277,7 @@ class checks that the lock file is a regular file with restricted
     effective.
 
     """
-    def __init__(self, run_path):
```
Comment (dannosliwcd) [inline geopmdpy/geopmdpy/system_files.py:1280]: The class-level and function-level docstrings need to be updated

**Review by dannosliwcd — COMMENTED** [review]
Changes look good overall.

If users see "requested control of application profile, but the geopm service already has an application profile controlling client", what should they do?

**Review by bgeltz — DISMISSED** [review]

**Review by bgeltz — APPROVED** [review]

**Pull request merged**